src/main/java/ch/njol/skript/expressions/ExprBookPages.java

```diff
@@ -21,94 +21,161 @@
-				"%itemtypes%'s [book] (pages|content)",
-				"[book] page %number% of %itemtypes%", 
-				"%itemtypes%'s [book] page %number%");
+			"[(all [[of] the]|the)] [book] (pages|content) of %itemtypes%",
+			"%itemtypes%'[s] [book] (pages|content)",
+			"[book] page %number% of %itemtypes%",
+			"%itemtypes%'[s] [book] page %number%");
 	}
-	
-	private static final ItemType bookItem = Aliases.javaItemType("book with text");
-	
-	@SuppressWarnings("null")
-	private Expression<ItemType> book;
+
 	@Nullable
 	private Expression<Number> page;
-	
+
 	@SuppressWarnings("null")
 	@Nullable
 	@Override
-	protected String[] get(Event e) {
-		ItemStack itemStack = book.getSingle(e).getRandom();
-		if (itemStack == null || !bookItem.isOfType(itemStack))
-			return null;
-		List<String> pages = ((BookMeta) itemStack.getItemMeta()).getPages();
-		if (page != null){
-			Number pageNumber = page.getSingle(e);
-			if (pageNumber == null){
-				return null;
+	protected String[] get(Event e, ItemType[] source) {
+		List<String> pages = new ArrayList<>();
+		for (ItemType itemType : source) {
+			ItemMeta meta = itemType.getItemMeta();
+			if (!(meta instanceof BookMeta))
+				continue;
+			BookMeta bookMeta = (BookMeta) meta;
+			if (page == null) {
+				pages.addAll(bookMeta.getPages());
+			} else {
+				Number pageNumber = page.getSingle(e);
+				if (pageNumber == null)
+					continue;
+				int page = pageNumber.intValue();
+				if (page <= 0 || page > bookMeta.getPageCount())
+					continue;
+				pages.add(bookMeta.getPage(page));
 			}
-			int page = pageNumber.intValue();
-			if ((page) > pages.size() || page < 1){
+		}
+		return pages.toArray(new String[0]);
+	}
+
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case SET:
+			case RESET:
+			case DELETE:
+				return CollectionUtils.array(page == null ? String[].class : String.class);
+			case ADD:
+				return page == null ? CollectionUtils.array(String[].class) : null;
+			default:
 				return null;
-			}
-			return new String[]{pages.get(page - 1)};
-		}else{
-			return pages.toArray(new String[pages.size()]);
 		}
 	}
-	
+
 	@Override
-	public boolean isSingle() {
-		return page != null;
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		List<ItemType> itemTypes = new ArrayList<>(Arrays.asList(getExpr().getArray(e)));
+		Number pageNumber = page == null ? -1 : page.getSingle(e);
+		if (pageNumber == null) return;
+		int page = pageNumber.intValue();
+		String[] newPages = delta == null ? null : new String[delta.length];
+		if (newPages != null)
+			for (int i = 0; i < delta.length; i++)
+				newPages[i] = delta[i] + "";
+		for (ItemType itemType : itemTypes) {
+			if (itemType == null) return;
+			ItemMeta meta = itemType.getItemMeta();
+			if (!(meta instanceof BookMeta)) return;
+			BookMeta bookMeta = (BookMeta) meta;
+			List<String> currentPages = new ArrayList<>(bookMeta.getPages());
```

Comment:
getPages already returns a List, and it can't be null, but empty.
```suggestion
			List<String> currentPages = bookMeta.getPages();
```
I also suggest pages as the variable name

Reply:
getPages() returns an immutable list, therefore I can't modify it. Which is why I wrap it in a new array list